### ISSUE TITLE
 Recurrent crash when using cuda_ad_rgb after a certain numbe of renders

### ISSUE THREAD

**Issue by osylum** [mitsuba-renderer/mitsuba3#578]
## Summary

 Recurrent crash when using cuda_ad_rgb after a certain numbe of renders

## System configuration

  OS: Windows-10
  CPU: Intel64 Family 6 Model 165 Stepping 5, GenuineIntel
  GPU: NVIDIA RTX A4000
  Python: 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)]
  NVidia driver: 517.40
  CUDA: 10.0.130
  LLVM: 15.-1.-1

  Dr.Jit: 0.4.0
  Mitsuba: 3.2.0
     Is custom build? False
     Compiled with: MSVC 19.34.31937.0
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

System information:

  OS: ...
  CPU: ...
  GPU: ...
  Python version: ...
  LLVM version: ...
  CUDA version: ...
  NVidia driver: ...

  Dr.Jit version: ...
  Mitsuba version: ...
     Compiled with: ...
     Variants compiled: ...

## Description

Program consistently stops/crash when using cuda_ad_rgb with error:

Critical Dr.Jit compiler failure: jit_var(r11793829): unknown variable!

It happens for very different codes. Tried also on another machine with same result. Below, it crashed at iteration 427 when running the caustics optimization code. It does not necessarily crash at this iteration, but systematically crashes after a number of renders.

Thank you

## Steps to reproduce

<pre><code>
import os,sys
from os.path import realpath, join
import importlib
import gc
import shutil

from timeit import default_timer as timer

import mitsuba as mi
print('available mitsuba variants:', mi.variants())
#mi.set_variant("scalar_rgb")
#mi.set_variant('llvm_ad_rgb') # could not find LLVM-C.dll
mi.set_variant('cuda_ad_rgb')
import drjit as dr

import matplotlib.pyplot as plt
import matplotlib.cm as cm

def caustics_optimization(mitsuba_scenes, output_dir):
    scene_filepath = os.path.join(mitsuba_scenes, 'cbox.xml')

    SCENE_DIR = realpath('../../../../RND_Relighting_Data/Mitsuba3/scenes')

    # config for image to display and emitter
    # As explained previously, depending on whether we are trying to reproduce a grayscale or colorful target image,
    # we setup the emitter to either emit constant white light or an RGB Bayer pattern.
    CONFIGS = {
        'wave': {
            'emitter': 'gray',
            'reference': join(SCENE_DIR, 'references/wave-1024.jpg'),
        },
        'sunday': {
            'emitter': 'bayer',
            'reference': join(SCENE_DIR, 'references/sunday-512.jpg'),
        },
    }

    # Pick one of the available configs
    config_name = 'sunday'
    # config_name = 'wave'

    config = CONFIGS[config_name]
    print('[i] Reference image selected:', config['reference'])
    plt.imshow(mi.Bitmap(config['reference']))
    plt.show()

    if 'PYTEST_CURRENT_TEST' not in os.environ:
        print('PYTEST_CURRENT_TEST not in os.environ')
        config.update({
            'render_resolution': (128, 128),
            'heightmap_resolution': (512, 512),
            'n_upsampling_steps': 4,
            'spp': 32,
            'max_iterations': 1000,
            'learning_rate': 3e-5,
        })
    else:
        # IGNORE THIS: When running under pytest, adjust parameters to reduce computation time
        config.update({
            'render_resolution': (64, 64),
            'heightmap_resolution': (128, 128),
            'n_upsampling_steps': 0,
            'spp': 8,
            'max_iterations': 25,
            'learning_rate': 3e-5,
        })

    output_dir = realpath(join(output_dir, config_name))
    os.makedirs(output_dir, exist_ok=True)
    print('[i] Results will be saved to:', output_dir)

    # Depending on the chosen configuration, a different type of emitter will need to be used. For this reason, we define the scene dynamically
    # Make sure that resources from the scene directory can be found
    print('file_resolver')
    mi.Thread.thread().file_resolver().append(SCENE_DIR)

    # helper to create a grid mesh
    def create_flat_lens_mesh(resolution):
        # Generate UV coordinates
        U, V = dr.meshgrid(
            dr.linspace(mi.Float, 0, 1, resolution[0]),
            dr.linspace(mi.Float, 0, 1, resolution[1]),
            indexing='ij'
        )
        texcoords = mi.Vector2f(U, V)

        # Generate vertex coordinates
        X = 2.0 * (U - 0.5)
        Y = 2.0 * (V - 0.5)
        vertices = mi.Vector3f(X, Y, 0.0)

        # Create two triangles per grid cell
        faces_x, faces_y, faces_z = [], [], []
        for i in range(resolution[0] - 1):
            for j in range(resolution[1] - 1):
                v00 = i * resolution[1] + j
                v01 = v00 + 1
                v10 = (i + 1) * resolution[1] + j
                v11 = v10 + 1
                faces_x.extend([v00, v01])
                faces_y.extend([v10, v10])
                faces_z.extend([v01, v11])

        # Assemble face buffer
        faces = mi.Vector3u(faces_x, faces_y, faces_z)

        # Instantiate the mesh object
        mesh = mi.Mesh("lens-mesh", resolution[0] * resolution[1], len(faces_x), has_vertex_texcoords=True)

        # Set its buffers
        mesh_params = mi.traverse(mesh)
        mesh_params['vertex_positions'] = dr.ravel(vertices)
        mesh_params['vertex_texcoords'] = dr.ravel(texcoords)
        mesh_params['faces'] = dr.ravel(faces)
        mesh_params.update()

        return mesh

    lens_res = config.get('lens_res', config['heightmap_resolution'])
    lens_fname = join(output_dir, 'lens_{}_{}.ply'.format(*lens_res))

    print('write mesh to file')
    if not os.path.isfile(lens_fname):
        m = create_flat_lens_mesh(lens_res)
        m.write_ply(lens_fname)
        print('[+] Wrote lens mesh ({}x{} tesselation) file to: {}'.format(*lens_res, lens_fname))

    # create emitter
    emitter = None
    if config['emitter'] == 'gray':
        emitter = {
            'type':'directionalarea',
            'radiance': {
                'type': 'spectrum',
                'value': 0.8
            },
        }
    elif config['emitter'] == 'bayer':
        # Bayer requires a bitmap texture
        bayer = dr.zeros(mi.TensorXf, (32, 32, 3))
        bayer[ ::2,  ::2, 2] = 2.2
        bayer[ ::2, 1::2, 1] = 2.2
        bayer[1::2, 1::2, 0] = 2.2

        emitter = {
            'type':'directionalarea',
            'radiance': {
                'type': 'bitmap',
                'bitmap': mi.Bitmap(bayer),
                'raw': True,
                'filter_type': 'nearest'
            },
        }

    # The chosen light source emits light in a single direction,
    # which would be very difficult (or impossible) to sample correctly with a standard path tracer.
    # For this reason, we use a particle tracer (ptracer), which starts rays from the emitters rather than the sensor.
    integrator = {
        'type': 'ptracer',
        'samples_per_pass': 256,
        'max_depth': 4,
        'hide_emitters': False,
    }

    # Looking at the receiving plane, not looking through the lens
    sensor_to_world = mi.ScalarTransform4f.look_at(
        target=[0, -20, 0],
        origin=[0, -4.65, 0],
        up=[0, 0, 1]
    )
    resx, resy = config['render_resolution']
    sensor = {
        'type': 'perspective',
        'near_clip': 1,
        'far_clip': 1000,
        'fov': 45,
        'to_world': sensor_to_world,

        'sampler': {
            'type': 'independent',
            'sample_count': 512  # Not really used
        },
        'film': {
            'type': 'hdrfilm',
            'width': resx,
            'height': resy,
            'pixel_format': 'rgb',
            'rfilter': {
                # Important: smooth reconstruction filter with a footprint larger than 1 pixel.
                'type': 'gaussian'
            }
        },
    }

    scene = {
        'type': 'scene',
        'sensor': sensor,
        'integrator': integrator,
        # Glass BSDF
        'simple-glass': {
            'type': 'dielectric',
            'id': 'simple-glass-bsdf',
            'ext_ior': 'air',
            'int_ior': 1.5,
            'specular_reflectance': { 'type': 'spectrum', 'value': 0 },
        },
        'white-bsdf': {
            'type': 'diffuse',
            'id': 'white-bsdf',
            'reflectance': { 'type': 'rgb', 'value': (1, 1, 1) },
        },
        'black-bsdf': {
            'type': 'diffuse',
            'id': 'black-bsdf',
            'reflectance': { 'type': 'spectrum', 'value': 0 },
        },
        # Receiving plane
        'receiving-plane': {
            'type': 'obj',
            'id': 'receiving-plane',
            'filename': 'meshes/rectangle.obj',
            'to_world': \
                mi.ScalarTransform4f.look_at(
                    target=[0, 1, 0],
                    origin=[0, -7, 0],
                    up=[0, 0, 1]
                ).scale((5, 5, 5)),
            'bsdf': {'type': 'ref', 'id': 'white-bsdf'},
        },
        # Glass slab, excluding the 'exit' face (added separately below)
        'slab': {
            'type': 'obj',
            'id': 'slab',
            'filename': 'meshes/slab.obj',
            'to_world': mi.ScalarTransform4f.rotate(axis=(1, 0, 0), angle=90),
            'bsdf': {'type': 'ref', 'id': 'simple-glass'},
        },
        # Glass rectangle, to be optimized
        'lens': {
            'type': 'ply',
            'id': 'lens',
            'filename': lens_fname,
            'to_world': mi.ScalarTransform4f.rotate(axis=(1, 0, 0), angle=90),
            'bsdf': {'type': 'ref', 'id': 'simple-glass'},
        },

        # Directional area emitter placed behind the glass slab
        'focused-emitter-shape': {
            'type': 'obj',
            'filename': 'meshes/rectangle.obj',
            'to_world': mi.ScalarTransform4f.look_at(
                target=[0, 0, 0],
                origin=[0, 5, 0],
                up=[0, 0, 1]
            ),
            'bsdf': {'type': 'ref', 'id': 'black-bsdf'},
            'focused-emitter': emitter,
        },
    }

    scene = mi.load_dict(scene)

    # load image, make sure resolution maps with sensor, return as TensorXf
    def load_ref_image(config, resolution, output_dir):
        b = mi.Bitmap(config['reference'])
        b = b.convert(mi.Bitmap.PixelFormat.RGB, mi.Bitmap.Float32, False)
        if b.size() != resolution:
            b = b.resample(resolution)

        mi.util.write_bitmap(join(output_dir, 'out_ref.exr'), b)

        print('[i] Loaded reference image from:', config['reference'])
        return mi.TensorXf(b)

    # Make sure the reference image will have a resolution matching the sensor
    print('load reference image')
    sensor = scene.sensors()[0]
    crop_size = sensor.film().crop_size()
    image_ref = load_ref_image(config, crop_size, output_dir=output_dir)

    # we optimize values of a high-resolution heightmap.
    # Here, we create the heightmap texture and create an optimizer that will work on its values.
    print('create heightmap texture')
    initial_heightmap_resolution = [r // (2 ** config['n_upsampling_steps'])
                                    for r in config['heightmap_resolution']]
    upsampling_steps = dr.sqr(dr.linspace(mi.Float, 0, 1, config['n_upsampling_steps']+1, endpoint=False).numpy()[1:])
    upsampling_steps = (config['max_iterations'] * upsampling_steps).astype(int)
    print('The resolution of the heightfield will be doubled at iterations:', upsampling_steps)

    heightmap_texture = mi.load_dict({
        'type': 'bitmap',
        'id': 'heightmap_texture',
        'bitmap': mi.Bitmap(dr.zeros(mi.TensorXf, initial_heightmap_resolution)),
        'raw': True,
    })

    # Actually optimized: the heightmap texture
    print('create optimizer')
    params = mi.traverse(heightmap_texture)
    params.keep(['data'])
    opt = mi.ad.Adam(lr=config['learning_rate'], params=params)

    #  the lens’ vertices will displaced from their original position along their normal by the value of the heightmap
    params_scene = mi.traverse(scene)

    # We will always apply displacements along the original normals and
    # starting from the original positions.
    print('create surface interactions')
    positions_initial = dr.unravel(mi.Vector3f, params_scene['lens.vertex_positions'])
    normals_initial   = dr.unravel(mi.Vector3f, params_scene['lens.vertex_normals'])

    lens_si = dr.zeros(mi.SurfaceInteraction3f, dr.width(positions_initial))
    lens_si.uv = dr.unravel(type(lens_si.uv), params_scene['lens.vertex_texcoords'])

    def apply_displacement(amplitude = 1.):
        # Enforce reasonable range. For reference, the receiving plane
        # is 7 scene units away from the lens.
        vmax = 1 / 100.
        params['data'] = dr.clamp(params['data'], -vmax, vmax)
        dr.enable_grad(params['data'])

        height_values = heightmap_texture.eval_1(lens_si)
        new_positions = (height_values * normals_initial * amplitude + positions_initial)
        params_scene['lens.vertex_positions'] = dr.ravel(new_positions)
        params_scene.update()

    # define loss function
    # We divide out the average brightness in the loss so that
    # the general brightness of the emitter (set arbitrarily) does not interfere with the optimization.
    def scale_independent_loss(image, ref):
        """Brightness-independent L2 loss function."""
        scaled_image = image / dr.mean(dr.detach(image))
        scaled_ref = ref / dr.mean(ref)
        return dr.mean(dr.sqr(scaled_image - scaled_ref))

    # We add two common tricks to improve the quality of the optimization:
    # - Increasing the rendering quality (sample count) and decreasing the learning rate towards the end of the optimization.
    # - Progressively increasing the resolution of the heightmap being optimized.
    import time
    start_time = time.time()
    mi.set_log_level(mi.LogLevel.Warn)
    iterations = config['max_iterations']
    loss_values = []
    spp = config['spp']

    # train
    print(f'start training for {iterations} iterations')
    for it in range(iterations):

        t0 = time.time()

        # Apply displacement and update the scene BHV accordingly
        apply_displacement()

        # Perform a differentiable rendering of the scene
        image = mi.render(scene, params, seed=it, spp=2 * spp, spp_grad=spp)

        # Scale-independent L2 function
        loss = scale_independent_loss(image, image_ref)

        # Back-propagate errors to input parameters and take an optimizer step
        dr.backward(loss)

        # Take a gradient step
        opt.step()

        # Increase resolution of the heightmap
        if it in upsampling_steps:
            opt['data'] = dr.upsample(opt['data'], scale_factor=(2, 2, 1))

        # Carry over the update to our "latent variable" (the heightmap values)
        params.update(opt)

        # Log progress
        elapsed_ms = 1000. * (time.time() - t0)
        current_loss = loss[0]
        loss_values.append(current_loss)
        mi.Thread.thread().logger().log_progress(
            it / (iterations-1),
            f'Iteration {it:03d}: loss={current_loss:g} (took {elapsed_ms:.0f}ms)',
            'Caustic Optimization', '')

        # Increase rendering quality toward the end of the optimization
        if it in (int(0.7 * iterations), int(0.9 * iterations)):
            spp *= 2
            opt.set_learning_rate(0.5 * opt.lr['data'])

    end_time = time.time()
    print(((end_time - start_time) * 1000) / iterations, ' ms per iteration on average')
    mi.set_log_level(mi.LogLevel.Info)

    # store results
    fname = join(output_dir, 'heightmap_final.exr')
    plt.imshow(mi.util.write_bitmap(fname, params['data']))
    plt.show()
    print('[+] Saved final heightmap state to:', fname)

    fname = join(output_dir, 'lens_displaced.ply')
    apply_displacement()
    lens_mesh = [m for m in scene.shapes() if m.id() == 'lens'][0]
    lens_mesh.write_ply(fname)
    print('[+] Saved displaced lens to:', fname)

    # show results
    def show_image(ax, img, title):
        ax.imshow(mi.util.convert_to_bitmap(img))
        ax.axis('off')
        ax.set_title(title)

    def show_heightmap(fig, ax, values, title):
        im = ax.imshow(values.squeeze(), vmax=1e-4)
        fig.colorbar(im, ax=ax)
        ax.axis('off')
        ax.set_title(title)

    fig, ax = plt.subplots(2, 2, figsize=(11, 10))
    ax = ax.ravel()
    ax[0].plot(loss_values)
    ax[0].set_xlabel('Iteration'); ax[0].set_ylabel('Loss value'); ax[0].set_title('Convergence plot')

    show_heightmap(fig, ax[1], params['data'].numpy(), 'Final heightmap')
    show_image(ax[2], image_ref, 'Reference')
    show_image(ax[3], image,     'Final state')
    plt.show()

if __name__ == "__main__":
    mitsuba_scenes = r'../scenes'
    output_dir = r'../outputs/caustics_optimization'
    caustics_optimization(mitsuba_scenes, output_dir)

</code></pre>

**Comment by njroussel** [comment]
Hi @osylum 

Could you try building from source on `master` and see if you're still able to replicate this bug?
